File: CalVisitor.py
from Gramatica_CalVisitor import Gramatica_CalVisitor
from Gramatica_CalParser import Gramatica_CalParser
import math

class CalVisitorPython(Gramatica_CalVisitor):
    # expr: MulDiv, AddSub, FactorialExpr, ToAtom
    def visitMulDiv(self, ctx: Gramatica_CalParser.MulDivContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == Gramatica_CalParser.MUL:
            return left * right
        else:
            return left / right

    def visitAddSub(self, ctx: Gramatica_CalParser.AddSubContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == Gramatica_CalParser.ADD:
            return left + right
        else:
            return left - right

    def visitFactorialExpr(self, ctx: Gramatica_CalParser.FactorialExprContext):
        n = int(self.visit(ctx.atom()))
        res = 1
        for i in range(1, n+1):
            res *= i
        return res

    def visitToAtom(self, ctx: Gramatica_CalParser.ToAtomContext):
        return self.visit(ctx.atom())

    # atom: UnaryMinus, Int, Double, Parens, SinExpr, CosExpr, TanExpr, SqrtExpr, LnExpr, LogExpr
    def visitUnaryMinus(self, ctx: Gramatica_CalParser.UnaryMinusContext):
        return -self.visit(ctx.atom())

    def visitInt(self, ctx: Gramatica_CalParser.IntContext):
        return int(ctx.INT().getText())

    def visitDouble(self, ctx: Gramatica_CalParser.DoubleContext):
        return float(ctx.DOUBLE().getText())

    def visitParens(self, ctx: Gramatica_CalParser.ParensContext):
        return self.visit(ctx.expr())

    def visitSinExpr(self, ctx: Gramatica_CalParser.SinExprContext):
        return math.sin(math.radians(self.visit(ctx.expr())))

    def visitCosExpr(self, ctx: Gramatica_CalParser.CosExprContext):
        return math.cos(math.radians(self.visit(ctx.expr())))

    def visitTanExpr(self, ctx: Gramatica_CalParser.TanExprContext):
        return math.tan(math.radians(self.visit(ctx.expr())))

    def visitSqrtExpr(self, ctx: Gramatica_CalParser.SqrtExprContext):
        return math.sqrt(self.visit(ctx.expr()))

    def visitLnExpr(self, ctx: Gramatica_CalParser.LnExprContext):
        return math.log(self.visit(ctx.expr()))

    def visitLogExpr(self, ctx: Gramatica_CalParser.LogExprContext):
        return math.log10(self.visit(ctx.expr()))

File: CalVisitor.java
import java.util.HashMap;
import java.util.Map;

public class CalVisitor extends Gramatica_CalBaseVisitor<Double> {
    private Map<String, Double> memory = new HashMap<>();
    private boolean grados = true; // true = grados, false = radianes

	
    
    /** Multiplicación / División */
    @Override
    public Double visitMulDiv(Gramatica_CalParser.MulDivContext ctx) {
        double left = visit(ctx.expr(0));
        double right = visit(ctx.expr(1));
        if (ctx.op.getType() == Gramatica_CalParser.MUL) return left * right;
        return left / right;
    }
    
    /** Imprimir: expr NEWLINE */
    @Override
    public Double visitPrintExpr(Gramatica_CalParser.PrintExprContext ctx) {
        Double value = visit(ctx.expr());
        System.out.println(value);
        return 0.0;
    }

    /** Entero */
    @Override
    public Double visitInt(Gramatica_CalParser.IntContext ctx) {
        return Double.valueOf(ctx.INT().getText());
    }

    /** Decimal */
    @Override
    public Double visitDouble(Gramatica_CalParser.DoubleContext ctx) {
        return Double.valueOf(ctx.DOUBLE().getText());
    }

    /** Paréntesis */
    @Override
    public Double visitParens(Gramatica_CalParser.ParensContext ctx) {
        return visit(ctx.expr());
    }

    /** Suma / Resta */
    @Override
    public Double visitAddSub(Gramatica_CalParser.AddSubContext ctx) {
        double left = visit(ctx.expr(0));
        double right = visit(ctx.expr(1));
        if (ctx.op.getType() == Gramatica_CalParser.ADD) return left + right;
        return left - right;
    }

    @Override
public Double visitFactorialExpr(Gramatica_CalParser.FactorialExprContext ctx) {
    double value = visit(ctx.atom());
    return (double) factorial((int) value);
}

private double factorial(int n) {
    if (n < 0) throw new RuntimeException("Factorial no definido para negativos");
    double result = 1;
    for (int i = 2; i <= n; i++) {
        result *= i;
    }
    return result;
}


    /** sin(expr) */
    @Override
    public Double visitSinExpr(Gramatica_CalParser.SinExprContext ctx) {
        double value = visit(ctx.expr());
        if (grados) value = Math.toRadians(value);
        return Math.sin(value);
    }

    /** cos(expr) */
    @Override
    public Double visitCosExpr(Gramatica_CalParser.CosExprContext ctx) {
        double value = visit(ctx.expr());
        if (grados) value = Math.toRadians(value);
        return Math.cos(value);
    }

    /** tan(expr) */
    @Override
    public Double visitTanExpr(Gramatica_CalParser.TanExprContext ctx) {
        double value = visit(ctx.expr());
        if (grados) value = Math.toRadians(value);
        return Math.tan(value);
    }

    /** sqrt(expr) */
    @Override
    public Double visitSqrtExpr(Gramatica_CalParser.SqrtExprContext ctx) {
        double value = visit(ctx.expr());
        return Math.sqrt(value);
    }

    /** ln(expr) */
    @Override
    public Double visitLnExpr(Gramatica_CalParser.LnExprContext ctx) {
        double value = visit(ctx.expr());
        return Math.log(value);
    }

    /** log(expr) */
    @Override
    public Double visitLogExpr(Gramatica_CalParser.LogExprContext ctx) {
        double value = visit(ctx.expr());
        return Math.log10(value);
    }

    /** Alternar grados/radianes */
    public void setgrados(boolean grados) {
        this.grados = grados;
    }
}
